File: leetcode/59.py
class Solution:
    def generateMatrix(self, n):
        if n == 0:
            return []  
        res = [[0] * n for x in range(n)]       
        num = 1      
        res[0][0] = 1
        for i in range((n + 1) / 2):
            for j in range(i, n - i):
                res[i][j] = num
                num += 1         
            for j in range(i + 1, n - i):
                res[j][n - i - 1] = num
                num += 1          
            for j in range(n - i - 2, i - 1, -1):
                res[n - i - 1][j] = num
                num += 1            
            for j in range(n - i - 2, i, -1):
                res[j][i] = num
                num += 1        
        return res

File: leetcode/59.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int> > ans(n, vector<int>(n, 0));
        int now = 0;
        int u = 0, d = n - 1, l = 0, r = n - 1;
        while (1) {
            for (int i = l; i <= r; ++i)    ans[u][i] = ++now;
            if (++u > d)    break;
            for (int i = u; i <= d; ++i)    ans[i][r] = ++now;
            if (--r < l)    break;
            for (int i = r; i >= l; --i)    ans[d][i] = ++now;
            if (--d < u)    break;
            for (int i = d; i >= u; --i)    ans[i][l] = ++now;
            if (++l > r)    break;
        }
        return ans;
    }
};